File: moo3-client/src/ui/ship/shipDesignRules.ts
import type { Key } from '../../i18n';
import type {
  ShipComponentOption,
  ShipDesign,
  ShipHull,
  WeaponModification,
} from '../../api/types';

/**
 * Правила проекта корабля на клиенте — п. 8.
 *
 * Повторяют `ShipDesignRules.java`, и это сознательный повтор: окно дизайна считает
 * проект на лету, пока игрок его собирает, а сохранённого проекта ещё нет — спросить
 * сервер не о чем. Правда о числах остаётся на сервере: он пересчитает всё при
 * сохранении и откажет, если проект не влезает.
 *
 * Единственное, чего здесь нет, — расовых процентов: их даёт `Fleet.attackPercent` и
 * `Fleet.defensePercent`, и они передаются параметром.
 */

/**
 * Компонент проекта: что взято, сколько раз и с какими модификациями — п. 8.
 *
 * Модификации бывают только у оружия: тяжёлое орудие, ближняя оборона, скорострельное,
 * бронебойное, обволакивающее. Они меняют урон, меткость, цену и место того ствола, на
 * который поставлены, — и одинаковые пушки с разными модификациями стоят в проекте
 * разными строками, как и в оригинале.
 */
export interface Pick {
  component: ShipComponentOption;
  count: number;
  modifications: WeaponModification[];
}

/** Сумма процентов модификаций одного вида: они складываются, как и на сервере. */
const modPercent = (pick: Pick, field: keyof WeaponModification): number =>
  pick.modifications.reduce((total, mod) => total + (mod[field] as number), 0);

/** Место одного такого ствола: модификация делает его крупнее или мельче. */
export const pickSpace = (pick: Pick, hull: ShipHull): number =>
  Math.max(1, Math.trunc((componentSpace(pick.component, hull)
    * Math.max(0, 100 + modPercent(pick, 'spacePercent'))) / 100));

/** Цена одного такого ствола — по тому же правилу, что и место. */
export const pickCost = (pick: Pick, hull: ShipHull): number =>
  Math.max(1, Math.trunc((componentCost(pick.component, hull)
    * Math.max(0, 100 + modPercent(pick, 'costPercent'))) / 100));

/** Урон одного выстрела этого ствола с поправкой модификаций. */
export const pickDamage = (pick: Pick): number => {
  const base = pick.component.effects.find((effect) => effect.type === 'WEAPON_DAMAGE');
  return Math.max(1, Math.trunc(((base?.amount ?? 0)
    * Math.max(0, 100 + modPercent(pick, 'damagePercent'))) / 100));
};

/**
 * Сколько выстрелов делает этот ствол за залп: автоматический огонь утраивает их число,
 * разделяющаяся боеголовка — учетверяет (п. 8).
 */
export const pickShots = (pick: Pick): number => {
  const base = pick.component.effects.find((effect) => effect.type === 'WEAPON_SHOTS');
  return Math.max(1, Math.trunc(((base?.amount ?? 0)
    * Math.max(0, 100 + modPercent(pick, 'shotsPercent'))) / 100));
};

/** Надбавка модификаций к меткости этого ствола, в процентных пунктах. */
export const pickAttackPercent = (pick: Pick): number => modPercent(pick, 'attackPercent');

/** Характеристики собираемого проекта — те же, что покажет сервер после сохранения. */
export interface Totals {
  space: number;
  spaceUsed: number;
  cost: number;
  structure: number;
  armour: number;
  shield: number;
  speed: number;
  /** Боевая скорость: число оригинала, поле боя делит его на четыре. */
  combatSpeed: number;
  attack: number;
  defense: number;
  /** Уклонение от ракет сверх обычной защиты: постановщик помех. */
  missileEvasion: number;
  troops: number;
  command: number;
}

/**
 * Место компонента на этом корпусе. Броня, щит, двигатель и приборы обслуживают весь
 * корабль, поэтому растут вместе с корпусом; пушка занимает своё место на любом.
 */
export function componentSpace(component: ShipComponentOption, hull: ShipHull): number {
  return component.scalesWithHull ? component.space * hull.systemFactor : component.space;
}

/** Цена компонента на этом корпусе — по тому же правилу, что и место. */
export function componentCost(component: ShipComponentOption, hull: ShipHull): number {
  return component.scalesWithHull ? component.cost * hull.systemFactor : component.cost;
}

/** Сумма одного действия по всему проекту — с учётом количества компонентов. */
function sum(picks: Pick[], type: string): number {
  return picks.reduce((total, pick) => {
    const amount = pick.component.effects
      .filter((effect) => effect.type === type)
      .reduce((value, effect) => value + effect.amount, 0);
    return total + amount * pick.count;
  }, 0);
}

/**
 * Во сколько раз обволакивающий удар сильнее обычного — то же число, что на сервере
 * (`ShipDesignRules.ENVELOPING_SIDES`): в бою он приходится на цель четыре раза.
 */
const ENVELOPING_SIDES = 4;

/** Обволакивает ли этот ствол цель — сам собой (плазма, Копьё новы) или модификацией ENV. */
const pickEnvelops = (pick: Pick): boolean =>
  pick.component.envelops === true || pick.modifications.some((mod) => mod.envelops);

/**
 * Залп проекта: урон всех стволов за заход, без приборов и расы.
 *
 * Обволакивающий удар входит вчетверо — правилом оригинала и сервера: он приходится на
 * цель четыре раза, по разу на каждую сторону.
 */
function salvo(picks: Pick[]): number {
  return picks
    .filter((pick) => pick.component.slot === 'WEAPON')
    .reduce((total, pick) => total + pickDamage(pick) * pickShots(pick) * pick.count
      * (pickEnvelops(pick) ? ENVELOPING_SIDES : 1), 0);
}

/**
 * Влезет ли ещё один такой компонент — п. 8.
 *
 * Место корабля не бывает отрицательным: в оригинале система, которой не хватает места,
 * просто не ставится. Считается это ТЕМ ЖЕ правилом, что и занятое место, — иначе запрет
 * и показанное число однажды разойдутся.
 *
 * Гнездо на одного (двигатель, броня, щит, компьютер, помехи) сравнивается с учётом того,
 * что прежний компонент из него уйдёт: поменять тяжёлую броню на лёгкую можно всегда.
 */
export function fitsMore(
  hull: ShipHull,
  picks: Pick[],
  component: ShipComponentOption,
  racePercent: { attack: number; defense: number },
): boolean {
  const now = totals(hull, picks, racePercent);
  const replaced = isSingle(component.slot)
    ? picks.filter((pick) => pick.component.slot === component.slot)
    : [];
  const freed = replaced.reduce((total, pick) => total + pickSpace(pick, hull) * pick.count, 0);
  const next = componentSpace(component, hull);
  // Место считается ПОСЛЕ добавления: модуль, меняющий вместимость (боевые отсеки), даёт
  // её сам себе, и без пересчёта он не влезал бы в корабль, который сам же и расширяет.
  const after = totals(hull, [...picks.filter((pick) => !replaced.includes(pick)),
    { component, count: 1, modifications: [] }], racePercent);
  return now.spaceUsed - freed + next <= after.space;
}

/** Характеристики проекта — те же формулы, что на сервере. */
export function totals(
  hull: ShipHull,
  picks: Pick[],
  racePercent: { attack: number; defense: number },
): Totals {
  const space = Math.trunc((hull.space * (100 + sum(picks, 'SPACE_PERCENT'))) / 100);
  // Место и цена берутся у строки состава, а не у компонента: модификация ствола меняет
  // и то и другое — п. 8.
  const spaceUsed = picks.reduce((total, pick) => total + pickSpace(pick, hull) * pick.count, 0);
  const cost = picks.reduce((total, pick) => total + pickCost(pick, hull) * pick.count, hull.cost);

  const armour = Math.trunc(
    (sum(picks, 'ARMOUR') * hull.systemFactor * (100 + sum(picks, 'ARMOUR_PERCENT'))) / 100,
  );

  return {
    space,
    spaceUsed,
    cost,
    structure: Math.trunc((hull.structure * (100 + sum(picks, 'STRUCTURE_PERCENT'))) / 100),
    armour,
    shield: sum(picks, 'SHIELD') * hull.systemFactor,
    speed: sum(picks, 'SPEED'),
    combatSpeed: sum(picks, 'COMBAT_SPEED'),
    attack: Math.trunc(
      (salvo(picks) * (100 + sum(picks, 'ATTACK_PERCENT') + racePercent.attack)) / 100,
    ),
    // Уклонение корпуса: по крупному кораблю попадают чаще — п. 8.
    defense: Math.trunc(
      ((100 - hull.hitChancePercent + sum(picks, 'DEFENSE')) * (100 + racePercent.defense)) / 100,
    ),
    missileEvasion: sum(picks, 'MISSILE_EVASION'),
    troops: sum(picks, 'TROOPS'),
    command: hull.command,
  };
}

/** Названия гнёзд по-русски. */
export const SLOT_NAMES: Record<string, Key> = {
  ENGINE: 'design.slotName.ENGINE',
  ARMOR: 'design.slotName.ARMOR',
  SHIELD: 'design.slotName.SHIELD',
  COMPUTER: 'design.slotName.COMPUTER',
  ECM: 'design.slotName.ECM',
  WEAPON: 'design.slotName.WEAPON',
  SPECIAL: 'design.slotName.SPECIAL',
};

/** В гнездо влезает один компонент: двигатель, броня, щит, компьютер, помехи. */
export function isSingle(slot: string): boolean {
  return slot !== 'WEAPON' && slot !== 'SPECIAL';
}

/** Готовый проект — обратно в состав для правки: ячейку переписывают, а не создают с нуля. */
export function picksOf(
  design: ShipDesign,
  components: ShipComponentOption[],
  modifications: WeaponModification[] = [],
): Pick[] {
  const byCode = new Map(components.map((component) => [component.code, component]));
  const modByCode = new Map(modifications.map((mod) => [mod.code, mod]));
  return design.components
    .map((part) => ({
      component: byCode.get(part.code),
      count: part.count,
      modifications: (part.modifications ?? [])
        .map((code) => modByCode.get(code))
        .filter((mod): mod is WeaponModification => mod !== undefined),
    }))
    .filter((pick): pick is Pick => pick.component !== undefined);
}

File: moo3-server/src/main/java/com/moo3/server/service/ShipDesignRules.java
package com.moo3.server.service;

import com.moo3.server.domain.RaceEffects;
import com.moo3.server.domain.ShipComponent;
import com.moo3.server.domain.WeaponModification;
import com.moo3.server.domain.ShipHull;
import com.moo3.server.domain.ShipStats;
import com.moo3.server.domain.enums.ShipComponentSlot;
import com.moo3.server.domain.enums.ShipEffectType;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

/**
 * Числовые правила проекта корабля — п. 8. Всё, что считается по корпусу и компонентам,
 * живёт здесь: экран дизайна, стройка колонии и бой берут числа отсюда и своих формул
 * не заводят.
 * <p>
 * <b>Что взято из MOO II.</b> Шанс попасть по кораблю зависит от размера корпуса
 * (20 % по фрегату, 100 % по Leviathan) — отсюда уклонение {@link ShipHull#evasion()}.
 * Командные очки корпусов (1 у фрегата, 6 у Leviathan) — оттуда же.
 * <p>
 * <b>Что реконструировано.</b> Точных таблиц места, цены и урона игра не публиковала, а
 * StrategyWiki и зеркала вики отвечают отказом (403 и 402). Поэтому лестницы чисел
 * держатся на трёх величинах, которые в проекте уже были (Mass Driver 6, Laser 4,
 * Fusion Beam 10 урона; фрегат 30 места за 16 единиц), и на порядке технологий MOO II.
 * Все числа лежат в {@code ship-components.json} и правятся без пересборки.
 * <p>
 * <b>Правила расчёта.</b> Броня, щит, двигатель и приборы обслуживают весь корабль,
 * поэтому их место и цена растут вместе с корпусом; пушка занимает своё место на любом
 * корпусе. Из-за этого крупный корпус несёт пропорционально больше огня — как в MOO II,
 * где ради этого и строят дредноуты.
 */
@Service
public class ShipDesignRules {

    /**
     * Во сколько раз обволакивающий удар (ENV) сильнее обычного — п. 8.
     * <p>
     * <b>Реконструкция под наш щит.</b> В MOO II у корабля четыре стороны со своими
     * щитами, и такой удар приходится на все четыре разом, каждая получает полную силу.
     * Щит здесь один на корабль, поэтому то же самое выражено множителем.
     * <p>
     * Стоит оно ЗДЕСЬ, а не в правилах боя: множитель нужен обоим — и бою
     * ({@code BattleRules.envelopingFactor}), и силе проекта ({@code salvo}), — а одно
     * число в двух местах однажды разойдётся.
     */
    public static final int ENVELOPING_SIDES = 4;

    /**
     * Ядерный двигатель — Power, уровень 1 («Nuclear Fission»). Без него кораблю нечем
     * лететь, поэтому кораблестроения у империи нет вовсе — п. 8.
     */
    public static final String DRIVE_TECH = "nuclear-drive";

    /**
     * Топливные элементы — Chemistry, уровень 1 («Chemistry»). Без них корабль не отойдёт
     * от родной звезды — см. {@link com.moo3.server.domain.enums.FuelTech}.
     */
    public static final String FUEL_TECH = "standard-fuel-cells";

    /**
     * Звёздная база — орбитальная верфь колонии. Код здания из
     * {@code resources/Buildings/buildings.json}.
     */
    public static final String STAR_BASE = "star-base";

    /**
     * Какого размера корпуса сходят со стапеля колонии без звёздной базы — п. 8.
     * <p>
     * В MOO II звёздная база и есть верфь: «a star dock capable of building ships larger
     * than destroyers». Фрегат и эсминец — два наименьших корпуса, их строит любая
     * колония; всё, что крупнее, требует базы на орбите.
     */
    public static final Integer HULL_SIZE_WITHOUT_STAR_BASE = 2;

    /**
     * Умеет ли империя строить корабли — п. 8.
     * <p>
     * Нужны базовые уровни двух разделов: Power даёт двигатель, Chemistry — топливо. Оба
     * уровня общие: изучив уровень, империя получает все его технологии разом, поэтому
     * достаточно спросить по одной из каждого. До них нет ни окна дизайна, ни корабля
     * в списке стройки колонии — строить было бы не из чего.
     */
    public Boolean shipbuildingAvailable(Set<String> technologies) {
        return technologies.contains(DRIVE_TECH) && technologies.contains(FUEL_TECH);
    }

    /** Чего не хватает для кораблестроения — строкой для игрока; пусто, если хватает. */
    public List<String> missingShipbuildingTechnologies(Set<String> technologies) {
        List<String> missing = new ArrayList<>(2);
        if (!technologies.contains(DRIVE_TECH)) {
            missing.add("Nuclear Fission (Power)");
        }
        if (!technologies.contains(FUEL_TECH)) {
            missing.add("Chemistry (Chemistry)");
        }
        return missing;
    }

    /**
     * Наибольший размер корпуса, который поднимет эта колония, — п. 8: со звёздной базой
     * любой, без неё только два наименьших.
     */
    public Integer maxHullSize(Boolean hasStarBase) {
        return Boolean.TRUE.equals(hasStarBase) ? Integer.MAX_VALUE : HULL_SIZE_WITHOUT_STAR_BASE;
    }

    /**
     * Проект и его состав: компонент и сколько раз он взят. Пушек в проекте бывает
     * несколько одинаковых, поэтому состав — не множество, а пары с количеством.
     */
    public record Item(ShipComponent component, Integer count,
                       /** Модификации этого ствола — п. 8; у прочих гнёзд пусто. */
                       List<WeaponModification> modifications) {

        /** Состав без модификаций — так собираются гнёзда, где их не бывает. */
        public Item(ShipComponent component, Integer count) {
            this(component, count, List.of());
        }

        /** Место одного такого ствола на корпусе — с поправкой модификаций. */
        public Integer spaceOn(ShipHull hull) {
            return scaled(component.spaceOn(hull), WeaponModification::spacePercent);
        }

        /** Цена одного такого ствола — по тому же правилу, что и место. */
        public Integer costOn(ShipHull hull) {
            return scaled(component.costOn(hull), WeaponModification::costPercent);
        }

        /** Урон одного выстрела с поправкой модификаций. */
        public Integer damage() {
            return scaled(component.amount(ShipEffectType.WEAPON_DAMAGE),
                    WeaponModification::damagePercent);
        }

        /**
         * Выстрелов в залпе с поправкой модификаций — п. 8: разделяющаяся ракета (MIRV)
         * несёт четыре боеголовки, и каждая летит своей целью.
         */
        public Integer shots() {
            return scaled(component.amount(ShipEffectType.WEAPON_SHOTS),
                    WeaponModification::shotsPercent);
        }

        /** Насколько крепче сама ракета: настолько же труднее сбить её ближней обороной. */
        public Integer missileArmourPercent() {
            return modifications.stream()
                    .mapToInt(WeaponModification::missileArmourPercent)
                    .sum();
        }

        /** Насколько ракета быстрее обычной — перехватить её тем труднее. */
        public Integer missileSpeed() {
            return modifications.stream().mapToInt(WeaponModification::missileSpeed).sum();
        }

        /**
         * Дальность этого ствола в клетках — п. 8: тяжёлая установка бьёт вдвое дальше,
         * ближняя оборона — вдвое ближе.
         */
        public Integer rangeFrom(Integer base) {
            int total = 100 + modifications.stream()
                    .mapToInt(WeaponModification::rangePercent)
                    .sum();
            return Math.max(1, base * Math.max(0, total) / 100);
        }

        /**
         * Охватывает ли удар цель со всех сторон разом — ENV или сам ствол.
         * <p>
         * Второе — правило оригинала: плазменная пушка и Копьё новы обволакивают цель сами
         * собой, и модификации ENV им не предлагают (`ship-components.json`, признак
         * {@code envelops}).
         */
        public Boolean envelops() {
            return Boolean.TRUE.equals(component.envelops())
                    || modifications.stream().anyMatch(m -> Boolean.TRUE.equals(m.envelops()));
        }

        /** Не слабеет ли удар с расстоянием — NR. */
        public Boolean noRangePenalty() {
            return modifications.stream().anyMatch(m -> Boolean.TRUE.equals(m.noRangePenalty()));
        }

        /** Срезает ли ракета уклонение цели вдвое — помехозащита ECCM. */
        public Boolean halvesEvasion() {
            return modifications.stream().anyMatch(m -> Boolean.TRUE.equals(m.halvesEvasion()));
        }

        /** Бьёт ли ракета по двигателю цели, пробившись сквозь щит, — наведение EMG. */
        public Boolean hitsEngine() {
            return modifications.stream().anyMatch(m -> Boolean.TRUE.equals(m.hitsEngine()));
        }

        /** Надбавка модификаций к меткости этого ствола, в процентных пунктах. */
        public Integer attackPercent() {
            return modifications.stream()
                    .mapToInt(WeaponModification::attackPercent)
                    .sum();
        }

        /** Пробивает ли выстрел броню насквозь. */
        public Boolean piercesArmour() {
            return modifications.stream().anyMatch(m -> Boolean.TRUE.equals(m.piercesArmour()));
        }

        /** Держит ли щит такой выстрел. */
        public Boolean piercesShield() {
            return modifications.stream().anyMatch(m -> Boolean.TRUE.equals(m.piercesShield()));
        }

        /**
         * Проценты модификаций складываются, а не перемножаются: в MOO II их и берут по
         * одной-две, а сложение оставляет числа круглыми — тяжёлое скорострельное стоит
         * ровно вдвое с половиной.
         */
        private Integer scaled(Integer base, java.util.function.ToIntFunction<WeaponModification> percent) {
            int total = 100 + modifications.stream().mapToInt(percent).sum();
            return Math.max(1, base * Math.max(0, total) / 100);
        }
    }

    /**
     * Считает характеристики проекта.
     *
     * @param race расовые проценты атаки и защиты кораблей — п. 7; они применяются
     *             последними, поверх приборов
     */
    public ShipStats stats(ShipHull hull, List<Item> items, RaceEffects race) {
        int spacePercent = sum(items, ShipEffectType.SPACE_PERCENT);
        int space = hull.space() * (100 + spacePercent) / 100;

        int spaceUsed = 0;
        int cost = hull.cost();
        for (Item item : items) {
            // Место и цена берутся у состава, а не у компонента: модификация ствола
            // меняет и то и другое — п. 8.
            spaceUsed += item.spaceOn(hull) * item.count();
            cost += item.costOn(hull) * item.count();
        }
        // Феодализм строит корабли на треть дешевле — п. 14. Скидка идёт от готовой цены
        // корпуса с приборами, а не от корпуса: в MOO II дешевле обходится весь корабль,
        // и это касается не только боевых — колониальные и грузовые тоже.
        cost = Math.max(1, cost * (100 + race.shipCostPercent()) / 100);

        int structure = hull.structure() * (100 + sum(items, ShipEffectType.STRUCTURE_PERCENT)) / 100;

        // Броня и щиты покрывают весь корпус, поэтому очки множатся на его размер —
        // тем же множителем, что место и цена. Тяжёлая броня добавляется процентом сверху.
        int armour = sum(items, ShipEffectType.ARMOUR) * hull.systemFactor()
                * (100 + sum(items, ShipEffectType.ARMOUR_PERCENT)) / 100;
        int shield = sum(items, ShipEffectType.SHIELD) * hull.systemFactor();

        int attack = salvo(items) * (100 + sum(items, ShipEffectType.ATTACK_PERCENT)
                + race.shipAttackPercent()) / 100;
        int defense = (hull.evasion() + sum(items, ShipEffectType.DEFENSE))
                * (100 + race.shipDefensePercent()) / 100;

        return new ShipStats(
                space,
                spaceUsed,
                cost,
                structure,
                armour,
                shield,
                sum(items, ShipEffectType.SPEED),
                sum(items, ShipEffectType.COMBAT_SPEED),
                attack,
                defense,
                sum(items, ShipEffectType.MISSILE_EVASION),
                sum(items, ShipEffectType.TROOPS),
                hull.command());
    }

    /**
     * Боевая сила одного корабля: чем он больнее бьёт, дольше живёт и труднее ловится, тем
     * больше весит в бою — п. 8.
     * <p>
     * Реконструкция: сравнивать флоты чем-то нужно, а быстрый бой ({@code
     * stub/SpaceBattleService}) поля не считает. Живучесть — прочность с бронёй и щитом;
     * деление держит числа в том же масштабе, в каком бой считал простые корабли; меньше
     * единицы сила не бывает — корабль на поле всегда что-то да значит.
     * <p>
     * <b>Защита входит тем же множителем, что и атака</b> — решение хозяина проекта, и
     * вот на чём оно стоит. Раньше её в силе не было вовсе: сила была «урон на живучесть»,
     * и сторона расы «пилоты» не меняла НИЧЕГО ни в мериле баланса, ни в исходе боя —
     * партии ИИ решаются только быстрым боём, а он сравнивает ровно эту силу. Измерено:
     * хорошие пилоты -0,24, отличные -0,17 по военному мерилу, то есть ноль. Уклонение
     * при этом в игре есть и работает — но лишь в тактическом бою, в кривой попадания,
     * которой партии ИИ не видят.
     * <p>
     * <b>Почему {@code 100 + защита}, а не голая защита.</b> Уклонение корпуса считается
     * как {@code 100 − шанс попадания}, и у Leviathan, звёздной крепости и наземных
     * батарей оно РАВНО НУЛЮ: по ним попадают всегда. Голый множитель обнулил бы их силу
     * вовсе, чего быть не должно — их защита это броня и щиты, а не верткость. Сотня и
     * есть та база «по мне всегда попадают», от которой уклонение считается прибавкой:
     * фрегат (уклонение 80) весит в 1,8 раза больше, чем весил бы неповоротливым, а
     * звёздная крепость — ровно столько же, сколько и раньше.
     */
    public Integer power(ShipStats stats) {
        int endurance = stats.structure() + stats.armour() + stats.shield();
        long evasion = 100L + Math.max(0, stats.defense());
        return (int) Math.max(1, (long) stats.attack() * endurance * evasion / 10_000L);
    }

    /**
     * Залп проекта: урон всех стволов за один заход, без приборов и расы. Урон берётся у
     * состава — модификация ствола его меняет (п. 8).
     */
    private Integer salvo(List<Item> items) {
        int salvo = 0;
        // Обволакивающий удар входит в залп вчетверо: в бою он и приходится на цель
        // четыре раза (ENVELOPING_SIDES). Без этого сила проекта расходилась бы с боем —
        // а её читают ИИ, сравнение флотов и прибор балансировки.
        for (Item item : items) {
            if (item.component().slot() == ShipComponentSlot.WEAPON) {
                int sides = Boolean.TRUE.equals(item.envelops()) ? ENVELOPING_SIDES : 1;
                salvo += item.damage() * item.shots() * item.count() * sides;
            }
        }
        return salvo;
    }

    /** Сумма одного эффекта по всему проекту — с учётом количества компонентов. */
    private Integer sum(List<Item> items, ShipEffectType type) {
        int total = 0;
        for (Item item : items) {
            total += item.component().amount(type) * item.count();
        }
        return total;
    }
}
